### ISSUE TITLE
installation blues...

### ISSUE THREAD

**Issue by houndofsound** [ocaml/opam#3506]
so from a fresh install:
progman@farm:~$ opam init
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.
```
[WARNING] Recommended external solver aspcud not found.
[ERROR] No OCaml compiler found in path. You should use:

            opam init --comp=VERSION

[ERROR] Initialisation failed
OpamGlobals.Exit(66)
```
why does it want ocaml right now?
so I take a shot at giving it a version...
```
progman@farm:~$ opam init --comp=4.07
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.

[WARNING] Recommended external solver aspcud not found.

=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[default] synchronized from https://opam.ocaml.org
Cannot find /home/farmerc/.opam/compilers/4.07/4.07/4.07.comp: 4.07 is not a valid compiler name.
[ERROR] Initialisation failed
OpamGlobals.Exit(1)
farmerc@cfarm:~$ opam init --comp=4.07
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.

[WARNING] Recommended external solver aspcud not found.

=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[default] synchronized from https://opam.ocaml.org
Cannot find /home/farmerc/.opam/compilers/4.07/4.07/4.07.comp: 4.07 is not a valid compiler name.
[ERROR] Initialisation failed
OpamGlobals.Exit(1)
```
luckily I am familiar with the EXACT versioning of ocaml rather than a complete NUBE ....
```
farmerc@cfarm:~$ opam init --comp=4.07.0
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.

[WARNING] Recommended external solver aspcud not found.

=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[default] synchronized from https://opam.ocaml.org

=-=- Installing compiler 4.07.0 -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[compiler.get] https://github.com/ocaml/ocaml/archive/4.07.0.tar.gz downloaded
Now compiling OCaml. This may take a while, please bear with us...
[ERROR] Compiler build failed at "make install":
        # opam-version 1.2.2
        # os           linux
        # command      make install
        # path         /home/farmerc/.opam/4.07.0/build/ocaml
        # exit-code    2
        # env-file     /home/farmerc/.opam/log/log-23032-8330f2.env
        # stdout-file  /home/farmerc/.opam/log/log-23032-8330f2.out
        # stderr-file  /home/farmerc/.opam/log/log-23032-8330f2.err
        ### stdout ###
        # do \
        # [...]
        #     install -m u=rwx,g=rwx,o=rx "$i.opt" "/home/farmerc/.opam/4.07.0/bin/$i.opt" && \
        #     (cd "/home/farmerc/.opam/4.07.0/bin/" && ln -sf "$i.opt" "$i"); \
        #   else \
        #     (cd "/home/farmerc/.opam/4.07.0/bin/" && ln -sf "$i.byte" "$i"); \
        #   fi; \
        # done
        # Makefile:234: recipe for target 'install' failed
        # make[1]: Leaving directory '/home/farmerc/.opam/4.07.0/build/ocaml/tools'
        # Makefile:600: recipe for target 'install' failed

[ERROR] Initialisation failed
OpamGlobals.Exit(66)
```
I don't know yet why it failed - the logs it purported to supply were gone.
I re-ran the command ... and it started the download of ocaml all over again.
not a virtuous loop.
must it insist on installing ocaml  (just a switch, no?) as part of its installation?

[rjbou] edit formatting

**Comment by dbuenzli** [comment]
Do you have `.` in you PATH by any chance ? If that's the case I suggest you remove it.

**Comment by houndofsound** [comment]
does the edit button work on this site? 
have tried 3 different browsers now!

**Comment by dra27** [comment]
This would be somewhat better asked on Discuss. OPAM 1.x has always needed ocaml at init to set up the system switch and compilers are always referred to by full version number. You could try out opam 2. You could also try 4.06.1 instead of 4.07.0 - by the look of it, the new installation mechanism is failing.

**Comment by houndofsound** [comment]
. NIMBY .

**Comment by houndofsound** [comment]
so the reason it requires ocaml on install is because .... it was always like that. great!
I post here because this is an issue that new users can be stonewalled by.

**Comment by dbuenzli** [comment]
It's a little bit difficult to follow you. How did you install `opam` in the first place.

**Comment by houndofsound** [comment]
I went to the install page on opam.ocaml.org
and it didn't mention being version 1.2.2
I will go find 2.0.0 ....

**Comment by dbuenzli** [comment]
Also note that you don't have an external solver installer, if you are using 1.2.2 this won't work on the current state of the `opam-repository`. Either install opam v2 or install an external solver like `apscud`.

**Comment by dra27** [comment]
@houndofsound wrote:
> . NIMBY .

No, Discuss is my backyard too. At the moment, *you* have problems with OPAM, not *OPAM* necessarily having problems itself - Discuss gives a wider audience for solving that. Issue trackers are then better when actual issues have been identified. It's also just possible that you've hit an issue with OCaml 4.07.0, which obviously won't ultimately want to be reported here.

> so the reason it requires ocaml on install is because .... it was always like that. great!

That's not what I wrote though, is it. I noted that it has always been like that because I happen to know you're not a new user, so assumed that you just hadn't run `opam init` in a  while (which is reasonable) and had forgotten - i.e. that you weren't reporting something new.

> I post here because this is an issue that new users can be stonewalled by.

That's noble, and obviously helpful, but you hadn't actually said that prior to that comment...

An OPAM 1.2.2 installation without the external solver is going to pretty useless (and is noted on the installation page, FWIW), as @dbuenzli notes - so more information on exactly how you installed really is necessary to move forwards.

**Comment by houndofsound** [comment]
my NIMBY was in relation to the dot in path question!

**Comment by AltGr** [comment]
> I moved .opam elsewhere for lack of space and linked to it and am able to create/delete in there.
but looks like bwrap doesn't like soft links.

can you try defining `OPAMROOT` to point to the target of the link? That would help debug if this is really a symlink issue or something else. Thanks!

**Comment by houndofsound** [comment]
after going to v2 I got errors on both 4.06.1 and 4.07.0 switch creation mentioning a missing file within the ocaml source - didn't look right so I went to the source under .opam and compiled it directly no problem.
so I gave up and with original 'out of space' error garnered, installed v1.2.2 instead
as this was meant to be a 5min jobby to run a script on a new machine.

I will go back in a while and redo the installs to both versions and document them.
I will answer your question here and put the v2 error (if it happens again) over on Discuss.
